desc_source.go

```diff
@@ -302,3 +294,79 @@ func addFilesToSet(allFiles []*descriptorpb.FileDescriptorProto, expanded map[st
 	}
 	return append(allFiles, fd.AsFileDescriptorProto())
 }
+
+// WriteProtoFiles will use the given descriptor source to resolve all the given
+// symbols and write proto files with their definitions to the given output directory.
+func WriteProtoFiles(outProtoDirPath string, descSource DescriptorSource, symbols ...string) error {
+	filenames, fds, err := getFileDescriptors(symbols, descSource)
+	if err != nil {
+		return err
+	}
+	// now expand that to include transitive dependencies in topologically sorted
+	// order (such that file always appears after its dependencies)
+	expandedFiles := make(map[string]struct{}, len(fds))
+	allFileDescriptors := make([]*desc.FileDescriptor, 0, len(fds))
+	for _, filename := range filenames {
+		allFileDescriptors = addFilesToFileDescriptorList(allFileDescriptors, expandedFiles, fds[filename])
+	}
+	pr := protoprint.Printer{}
+	// now we can serialize to files
+	for i := range allFileDescriptors {
+		if err := writeProtoFile(outProtoDirPath, allFileDescriptors[i], &pr); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func writeProtoFile(outProtoDirPath string, fd *desc.FileDescriptor, pr *protoprint.Printer) error {
+	fdFQName := fd.GetFullyQualifiedName()
+	dirPath := filepath.Dir(fdFQName)
+	outFilepath := filepath.Join(outProtoDirPath, dirPath)
+	if err := os.MkdirAll(outFilepath, 0755); err != nil {
+		return fmt.Errorf("failed to create directory %q: %v", outFilepath, err)
+	}
+	fileName := filepath.Base(fdFQName)
+	filePath := filepath.Join(outFilepath, fileName)
+
+	f, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to create proto file: %v", err)
+	}
+	defer f.Close()
+	if err := pr.PrintProtoFile(fd, f); err != nil {
+		return fmt.Errorf("failed to write proto file: %v", err)
+	}
+	return nil
+}
```

Comment:
Let's do one more pass on this, I have a few changes:
- simplify filepath calculations and naming
- improve errors with filename and error wrapping
- 0777 so the system level umask can take effect

```go
func writeProtoFile(outProtoDirPath string, fd *desc.FileDescriptor, pr *protoprint.Printer) error {
	outFile := filepath.Join(outProtoDirPath, fd.GetFullyQualifiedName())
	outDir := filepath.Dir(outFile)
	if err := os.MkdirAll(outDir, 0777); err != nil {
		return fmt.Errorf("failed to create directory %q: %w", outDir, err)
	}

	f, err := os.Create(outFile)
	if err != nil {
		return fmt.Errorf("failed to create proto file %q: %w", outFile, err)
	}
	defer f.Close()
	if err := pr.PrintProtoFile(fd, f); err != nil {
		return fmt.Errorf("failed to write proto file %q: %w", outFile, err)
	}
	return nil
}
```

Reply:
ready